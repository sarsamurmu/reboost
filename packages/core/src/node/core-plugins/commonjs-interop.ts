import * as t from '@babel/types';
import { NodePath } from '@babel/traverse';

import { ReboostPlugin } from '../index';

export const CommonJSInteropPlugin: ReboostPlugin = {
  name: 'core-commonjs-interop-plugin',
  transformAST(ast, { traverse }, filePath) {
    let program: NodePath<t.Program>;
    let importerIdentifier: t.Identifier;
    const replacements: [NodePath<t.ImportDeclaration>, t.ImportDeclaration, t.VariableDeclarator[]][] = [];

    traverse(ast, {
      Program(path) {
        program = path;
      },
      ImportDeclaration(path) {
        const declarators: t.VariableDeclarator[] = [];
        const identifier = path.scope.generateUidIdentifier('$import');
        if (!importerIdentifier) importerIdentifier = path.scope.generateUidIdentifier('$importer');

        path.node.specifiers.forEach((specifier) => {
          let usage;
          let importedName;
          const localName = specifier.local.name;
          const commons = [
            t.callExpression(
              t.identifier('__reboost_resolve'),
              [t.stringLiteral(path.node.source.value)]
            ),
            t.stringLiteral(filePath)
          ];

          if (t.isImportDefaultSpecifier(specifier)) {
            usage = 'Default';
          } else if (t.isImportNamespaceSpecifier(specifier)) {
            usage = 'All';
          } else if (t.isImportSpecifier(specifier)) {
            usage = 'Member';
            importedName = specifier.imported.name;
          }

          declarators.push(
            t.variableDeclarator(
              t.identifier(localName),
              t.callExpression(
                t.memberExpression(
                  importerIdentifier,
                  t.identifier(usage)
                ),
                importedName ? [
                  identifier,
                  t.stringLiteral(importedName),
                  ...commons
                ] : [identifier, ...commons]
              )
            )
          );
        });

        replacements.push([
          path,
          t.importDeclaration([
            t.importNamespaceSpecifier(identifier)
          ], t.stringLiteral(path.node.source.value)),
          declarators
        ]);
      }
    });

    if (importerIdentifier) {
      replacements.forEach(([path, replacement, declarators]) => {
        path.replaceWith(replacement);
        if (declarators.length) {
          const constDeclaration = t.variableDeclaration('const', declarators);
          path.insertAfter(constDeclaration);
        }
      });

      program.node.body.unshift(
        t.importDeclaration([
          t.importDefaultSpecifier(importerIdentifier)
        ], t.stringLiteral(`#/importer`))
      );
    }
  }
}
