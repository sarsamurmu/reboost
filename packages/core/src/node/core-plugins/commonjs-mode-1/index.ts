import { transformCommonJS } from './transform-commonjs';

import { ReboostPlugin } from '../../index';
import { uniqueID } from '../../utils';

export { transformCommonJS as runTransformation }

export const CommonJSMode1Plugin = (): ReboostPlugin => ({
  name: 'core-commonjs-mode-1-plugin',
  getCacheKey: () => 1,
  transformAST(ast, _, filePath) {
    transformCommonJS(ast, filePath, uniqueID(6));
  }
})
