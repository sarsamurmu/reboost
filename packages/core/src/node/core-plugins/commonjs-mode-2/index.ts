import anymatch from 'anymatch';

import { ReboostInstance, ReboostPlugin } from '../../index';
import { transformCommonJSToES6 } from './commonjs-to-es6';
import { uniqueID } from '../../utils';

export { transformCommonJSToES6 as runTransformation }

export const CommonJSMode2Plugin = (instance: ReboostInstance): ReboostPlugin => {
  const test = (file: string) => (
    !anymatch(instance.config.commonJSInterop.exclude, file) &&
    anymatch(instance.config.commonJSInterop.include, file)
  );

  return {
    name: 'core-commonjs-mode-2-plugin',
    getCacheKey: ({ serializeObject }) => serializeObject({
      include: instance.config.commonJSInterop.include,
      exclude: instance.config.commonJSInterop.exclude
    }),
    transformAST(ast, _, filePath) {
      if (test(filePath)) {
        transformCommonJSToES6(ast, uniqueID(6));
      }
    }
  }
}
