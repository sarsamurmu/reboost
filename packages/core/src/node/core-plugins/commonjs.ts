import * as t from '@babel/types';
import { NodePath } from '@babel/traverse';

import { builtinModules } from 'module';

import { ReboostPlugin } from '../index';
import { uniqueID } from '../utils';

export const isDeclared = (path: NodePath<any>, identifierName: string, findInProgram = true) => {
  const findInDeclarators = (id: t.LVal) => {
    let found = false;
    if (t.isIdentifier(id, { name: identifierName })) {
      found = true;
    } else if (t.isArrayPattern(id)) {
      id.elements.forEach((element) => {
        if (found) return;
        found = findInDeclarators(element);
      });
    } else if (t.isObjectPattern(id)) {
      id.properties.forEach((property) => {
        if (found) return;

        if (t.isRestElement(property)) {
          found = t.isIdentifier(property.argument, { name: identifierName });
        } else {
          found = findInDeclarators(property.value as any);
        }
      });
    }
    return found;
  }

  const findInParent = (astPath: NodePath<any>): boolean => {
    let found = false;

    const parent: NodePath<t.BlockStatement | t.Program> = astPath.findParent(
      (p) => p.isBlockStatement() || (findInProgram && p.isProgram())
    ) as any;
    if (parent) {
      parent.node.body.forEach((item) => {
        if (found) return;

        if (t.isVariableDeclaration(item)) {
          item.declarations.forEach((declaration) => {
            if (found) return;

            found = findInDeclarators(declaration.id);
          });
        }
      });

      if (!found) return findInParent(parent);
    }

    return found;
  }

  return findInParent(path);
}

export const CommonJSPlugin: ReboostPlugin = {
  name: 'core-commonjs-plugin',
  transformAST(ast, { traverse }) {
    let program: NodePath<t.Program>;
    let cjsModule = false;
    const uid = uniqueID(4);
    const modImports: t.ImportDeclaration[] = [];
    const importIdentifierMap = {} as Record<string, t.Identifier>;

    traverse(ast, {
      Program(path) {
        program = path;
      },
      CallExpression(path) {
        if (
          t.isIdentifier(path.node.callee, { name: 'require' }) &&
          t.isStringLiteral(path.node.arguments[0])
        ) {
          const importPath = path.node.arguments[0].value;
          const importIdentifier = importIdentifierMap[importPath] || path.scope.generateUidIdentifier(`$imported_${uid}`);

          // Don't resolve built-in modules like path, fs, etc.
          if (builtinModules.includes(importPath)) return;

          if (!(importPath in importIdentifierMap)) {
            importIdentifierMap[importPath] = importIdentifier;
            modImports.push(
              t.importDeclaration(
                [t.importNamespaceSpecifier(importIdentifier)],
                t.stringLiteral(importPath)
              )
            );
          }

          path.replaceWith(importIdentifier);
        }
      },
      MemberExpression(path) {
        if (cjsModule) return;
        if (
          t.isIdentifier(path.node.object, { name: 'module' }) &&
          t.isIdentifier(path.node.property, { name: 'exports' })
        ) {
          if (!isDeclared(path, 'module')) cjsModule = true;
        } else if (t.isIdentifier(path.node.object, { name: 'exports' })) {
          if (!isDeclared(path, 'exports')) cjsModule = true;
        }
      }
    });

    if (cjsModule) {
      const moduleExportsExp = t.memberExpression(t.identifier('module'), t.identifier('exports'));

      program.node.body.unshift(
        t.variableDeclaration(
          'const',
          [
            t.variableDeclarator(
              t.identifier('module'),
              t.objectExpression([
                t.objectProperty(
                  t.identifier('exports'),
                  t.objectExpression([])
                )
              ])
            ),
          ]
        ),
        t.variableDeclaration(
          'const',
          [
            t.variableDeclarator(
              t.identifier('exports'),
              moduleExportsExp
            )
          ]
        )
      );

      program.node.body.push(
        t.exportNamedDeclaration(
          t.variableDeclaration(
            'const',
            [t.variableDeclarator(
              t.identifier('__cjsExports'),
              moduleExportsExp
            )]
          )
        )
      );
    }

    program.node.body.unshift(...modImports);
  }
}
