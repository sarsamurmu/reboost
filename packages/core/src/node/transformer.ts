import { parse } from '@babel/parser';
import traverse, { NodePath } from '@babel/traverse';
import generate, { GeneratorOptions } from '@babel/generator';
import * as babelTypes from '@babel/types';
import { codeFrameColumns, SourceLocation } from '@babel/code-frame';
import anymatch from 'anymatch';
import { RawSourceMap, SourceMapConsumer } from 'source-map';
import chalk from 'chalk';
import MagicString from 'magic-string';

import fs from 'fs';
import path from 'path';

import { ReboostPlugin, PluginContext } from './index';
import { getConfig, getAddress } from './shared';
import { mergeSourceMaps, bind } from './utils';
import { resolveModule } from './core-plugins/resolver';

const fixPath = (pathString: string) => pathString.replace(/\\/g, '/');

const getCompatibleSourceMap = (map: RawSourceMap) => {
  const config = getConfig();

  map.sourceRoot = 'reboost:///';

  map.sources = map.sources.map((sourcePath: string) => {
    if (path.isAbsolute(sourcePath)) return fixPath(path.relative(config.rootDir, sourcePath));
    return fixPath(sourcePath);
  });

  map.sourcesContent = [];
  map.sources.forEach((sourcePath) => {
    const absolutePath = path.join(config.rootDir, sourcePath);
    if (fs.existsSync(absolutePath)) {
      map.sourcesContent.push(fs.readFileSync(absolutePath).toString());
    } else {
      console.log(chalk.red(`Unable to find file "${absolutePath}". Required for source map generation.`));
      map.sourcesContent.push(`Unable to find file in "${absolutePath}".`);
    }
  });

  map.file = undefined;

  return map;
}

const getPluginContext = (filePath: string, mergedDependencies: string[]): PluginContext => ({
  address: getAddress(),
  config: getConfig(),
  addDependency(dependency) {
    mergedDependencies.push(dependency);
  },
  chalk,
  getCompatibleSourceMap,
  MagicString,
  mergeSourceMaps,
  resolveModule
})

const handleError = ({ message }: Error, dependencies: string[]) => {
  console.log(chalk.red(message));

  return {
    code: `console.error(${JSON.stringify('[reboost] ' + message)})`,
    dependencies,
    error: true
  }
}

let pluginsInitiated = false;
let resolveHooks: ReboostPlugin['resolve'][];
let loadHooks: ReboostPlugin['load'][];
let transformContentHooks: ReboostPlugin['transformContent'][];
let transformIntoJSHooks: ReboostPlugin['transformIntoJS'][];
let transformASTHooks: ReboostPlugin['transformAST'][];

export const transformFile = async (filePath: string): Promise<{
  code: string;
  dependencies: string[];
  map?: string;
  imports?: string[];
  error?: boolean;
}> => {
  if (!pluginsInitiated) {
    const def = (a: any) => !!a;
    const plugins = getConfig().plugins.filter(def);

    resolveHooks = plugins.map((plugin) => plugin.resolve).filter(def);
    loadHooks = plugins.map((plugin) => plugin.load).filter(def);
    transformContentHooks = plugins.map((plugin) => plugin.transformContent).filter(def);
    transformIntoJSHooks = plugins.map((plugin) => plugin.transformIntoJS).filter(def);
    transformASTHooks = plugins.map((plugin) => plugin.transformAST).filter(def);
    
    pluginsInitiated = true;
  }

  let code: string;
  let sourceMap: RawSourceMap;
  let ast: babelTypes.Node;
  let type: string;
  let errorOccurred = false;
  const imports: string[] = [];
  const dependencies: string[] = [];
  const pluginContext = getPluginContext(filePath, dependencies);

  for (const hook of loadHooks) {
    let result = await bind(hook, pluginContext)(filePath);
    if (result) {
      ({ code } = result);
      ({ type } = result);
      if (result.map) sourceMap = result.map;
      break;
    }
  }

  for (const hook of transformContentHooks) {
    const result = await bind(hook, pluginContext)({ code, type, map: sourceMap }, filePath);
    if (result) {
      if (result instanceof Error) return handleError(result, dependencies);

      ({ code } = result);
      if (result.map) {
        // Here source maps sources can be null, like when source map is generated using MagicString (npm package)
        result.map.sources = result.map.sources.map((sourcePath) => !sourcePath ? filePath : sourcePath);
        sourceMap = sourceMap ? await mergeSourceMaps(sourceMap, result.map) : result.map;
      }
      if (result.type) ({ type } = result);
    }
  }

  for (const hook of transformIntoJSHooks) {
    const result = await bind(hook, pluginContext)({
      code,
      type,
      map: sourceMap
    }, filePath);

    if (result) {
      if (result instanceof Error) return handleError(result, dependencies);

      ({ code } = result);
      sourceMap = result.inputMap;
      type = 'js';
      break;
    }
  }

  if (type !== 'js') {
    console.log(chalk.red(`File with type "${type}" is not supported. You may need proper loader to transform this kind of files into JS.`));
  }

  try {
    ast = parse(code, {
      sourceType: 'module'
    });
  } catch (e) {
    let message = '';
    let consoleMessage = '';
    let frameMessage = e.message.replace(/\s*\(.*\)$/, '');
    let rawCode = code;
    let location: SourceLocation['start'] = e.loc;
    let unableToLocateFile = false;
    message += `Error while parsing "${filePath}"\n`;
    message += 'You may need proper loader to handle this kind of files.\n\n';

    consoleMessage += chalk.red(message);

    if (sourceMap) {
      const consumer = await new SourceMapConsumer(sourceMap);
      const originalLoc = consumer.originalPositionFor(e.loc);
      if (originalLoc.source) {
        const originalCode = consumer.sourceContentFor(originalLoc.source);
        if (originalCode) {
          rawCode = originalCode;
          location = originalLoc;
          location.column = location.column || 1;
        } else {
          const absPathToSource = path.join(getConfig().rootDir, originalLoc.source);
          if (fs.existsSync(absPathToSource)) {
            rawCode = fs.readFileSync(absPathToSource).toString();
            location = originalLoc;
            location.column = location.column || 1;
          } else {
            unableToLocateFile = true;
          }
        }
      }
    }

    if (unableToLocateFile) {
      let unableToLocateMsg = 'We are unable to locate the original file. ';
      unableToLocateMsg += 'This is not accurate, but it may help you at some point.\n\n';
      
      message += unableToLocateMsg + codeFrameColumns(code, { start: e.loc }, {
        message: frameMessage
      });

      consoleMessage += unableToLocateMsg + codeFrameColumns(code, { start: e.loc }, {
        highlightCode: true,
        message: frameMessage
      });
    } else {
      message += codeFrameColumns(rawCode, { start: location }, {
        message: frameMessage
      });

      consoleMessage += codeFrameColumns(rawCode, { start: location }, {
        highlightCode: true,
        message: frameMessage
      });
    }

    console.log(consoleMessage);

    return {
      code: `console.error('[reboost] ' + ${JSON.stringify(message)});`,
      dependencies,
      error: true
    }
  }

  for (const hook of transformASTHooks) await bind(hook, pluginContext)(ast, { traverse, types: babelTypes }, filePath);

  const resolvedPathMap = {} as Record<string, any>;
  const resolvePath = async (source: string, filePath: string) => {
    const keyStr = `${filePath}||${source}`;
    if (resolvedPathMap[keyStr]) return resolvedPathMap[keyStr];
    for (const hook of resolveHooks) {
      const resolvedPath = await hook(source, filePath);
      if (resolvedPath) {
        resolvedPathMap[keyStr] = resolvedPath;
        return resolvedPath
      }
    }
    console.log(chalk.red(`[reboost] Unable to resolve path "${source}" of "${filePath}"`));
    return null;
  }

  const resolveDeps = async (astPath: NodePath<babelTypes.ImportDeclaration> | NodePath<babelTypes.ExportDeclaration>) => {
    if ((astPath.node as any).source) {
      const source: string = (astPath.node as any).source.value;
      
      if (source === 'reboost/hmr') {
        (astPath.node as any).source.value = `/hmr?q=${encodeURI(filePath)}`;
      } else {
        let finalPath = null;
        let routed = false;
        if (source.startsWith('#/')) {
          finalPath = source.replace(/^#/, '');
          routed = true;
        } else {
          const resolvedPath = await resolvePath(source, filePath);
          if (resolvedPath) {
            if (resolvedPath.startsWith('#/')) {
              finalPath = resolvedPath.replace(/^#/, '');
              routed = true;
            } else {
              finalPath = resolvedPath;
              imports.push(resolvedPath);
            }
          } else {
            errorOccurred = true;
          }
        }

        (astPath.node as any).source.value = getAddress() + (routed
          ? encodeURI(finalPath)
          : finalPath
            ? `/transformed?q=${encodeURI(finalPath)}`
            : `/unresolved?import=${encodeURI(source)}`);
      }
    }
  }

  const promiseFunctions: (() => Promise<void>)[] = [];
  let astProgram: NodePath<babelTypes.Program>;

  traverse(ast, {
    Program(astPath) {
      astProgram = astPath;
    },
    ImportDeclaration(astPath) {
      promiseFunctions.push(async () => {
        await resolveDeps(astPath);
      });
      return false;
    },
    ExportDeclaration(astPath) {
      promiseFunctions.push(async () => {
        await resolveDeps(astPath);
      });
      return false;
    },
    CallExpression(astPath) {
      const t = babelTypes;
      if (t.isIdentifier(astPath.node.callee, { name: '__reboost_resolve' })) {
        promiseFunctions.push(async () => {
          astPath.replaceWith(
            t.stringLiteral(
              await resolvePath((astPath.node.arguments[0] as babelTypes.StringLiteral).value, filePath)
            )
          );
        });
      } else if (t.isImport(astPath.node.callee)) {
        // Rewrite dynamic imports
        const importDeclarations = astProgram.get('body').filter((p) => p.isImportDeclaration());
        let importerDeclaration = importDeclarations.find(
          (dec) => (dec as NodePath<babelTypes.ImportDeclaration>).node.source.value.includes('#/importer')
        ).node as babelTypes.ImportDeclaration;
        if (!importerDeclaration) {
          const identifier = astPath.scope.generateUidIdentifier('$importer');
          importerDeclaration = t.importDeclaration([
            t.importDefaultSpecifier(identifier)
          ], t.stringLiteral(`#/importer`));
          astProgram.node.body.unshift(importerDeclaration);
        }
        const importerIdentifierName = importerDeclaration.specifiers[0].local.name;
        astPath.replaceWith(
          t.callExpression(
            t.memberExpression(
              t.identifier(importerIdentifierName),
              t.identifier('Dynamic')
            ),
            [
              t.stringLiteral((astPath.node.arguments[0] as babelTypes.StringLiteral).value),
              t.stringLiteral(filePath)
            ]
          )
        );
      }
    }
  });

  for (const execute of promiseFunctions) await execute();

  const sourceMapsConfig = getConfig().sourceMaps;
  const sourceMapsEnabled = !anymatch(sourceMapsConfig.exclude, filePath) && anymatch(sourceMapsConfig.include, filePath);
  const { debugMode } = getConfig();
  let generatorOptions: GeneratorOptions = {
    sourceMaps: true,
    sourceFileName: fixPath(path.relative(getConfig().rootDir, filePath)),
    sourceRoot: 'reboost:///',
    minified: !debugMode
  }

  const { code: generatedCode, map: generatedMap } = generate(ast, sourceMapsEnabled ? generatorOptions : undefined);
  let map;

  if (sourceMap && sourceMapsEnabled) {
    const merged = await mergeSourceMaps(sourceMap, generatedMap);
    map = getCompatibleSourceMap(merged);
  }

  return {
    code: generatedCode,
    map: map && JSON.stringify(map, null, getConfig().debugMode ? 2 : 0),
    imports,
    dependencies,
    error: errorOccurred
  }
}
