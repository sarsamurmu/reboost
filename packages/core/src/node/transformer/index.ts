import generate, { GeneratorOptions } from '@babel/generator';
import anymatch from 'anymatch';
import { RawSourceMap } from 'source-map';
import chalk from 'chalk';
import MagicString from 'magic-string';

import fs from 'fs';
import path from 'path';

import { PluginContext } from '../index';
import { getConfig, getAddress } from '../shared';
import { mergeSourceMaps, toPosix } from '../utils';
import { resolve } from '../core-plugins/resolver';
import { process } from './processor';
import { resolveImports } from './import-resolver';

const getCompatibleSourceMap = (map: RawSourceMap) => {
  const config = getConfig();

  map.sourceRoot = 'reboost:///';

  map.sources = map.sources.map((sourcePath: string) => {
    if (path.isAbsolute(sourcePath)) return toPosix(path.relative(config.rootDir, sourcePath));
    return toPosix(sourcePath);
  });

  map.sourcesContent = [];
  map.sources.forEach((sourcePath) => {
    const absolutePath = path.join(config.rootDir, sourcePath);
    if (fs.existsSync(absolutePath)) {
      map.sourcesContent.push(fs.readFileSync(absolutePath).toString());
    } else {
      console.log(chalk.red(`Unable to find file "${toPosix(absolutePath)}". Required for source map generation.`));
      map.sourcesContent.push(`Unable to find file in "${toPosix(absolutePath)}".`);
    }
  });

  map.file = undefined;

  return map;
}

const getPluginContext = (filePath: string, mergedDependencies: string[]): PluginContext => ({
  address: getAddress(),
  config: getConfig(),
  addDependency(dependency) {
    mergedDependencies.push(toPosix(dependency));
  },
  chalk,
  getCompatibleSourceMap,
  getSourceMapComment(map) {
    let comment = '/*# sourceMappingURL=data:application/json;charset=utf-8;base64,';
    comment += Buffer.from(JSON.stringify(map)).toString('base64');
    comment += ' */';
    return comment;
  },
  MagicString,
  mergeSourceMaps,
  resolve
});

export const transformFile = async (filePath: string): Promise<{
  code: string;
  dependencies: string[];
  map?: string;
  imports?: string[];
  error?: boolean;
}> => {
  let errorOccurred = false;
  const imports: string[] = [];
  const dependencies: string[] = [];
  const pluginContext = getPluginContext(filePath, dependencies);

  const getErrorObj = (msg: string) => ({
    code: `console.error('[reboost] ' + ${JSON.stringify(msg)})`,
    error: true,
    dependencies
  });

  const processed = await process(filePath, pluginContext);

  if (processed.error) return getErrorObj(processed.error);

  const { ast, sourceMap } = processed;

  errorOccurred = await resolveImports(ast, filePath, imports);

  const sourceMapsConfig = getConfig().sourceMaps;
  const sourceMapsEnabled = !anymatch(sourceMapsConfig.exclude, filePath) && anymatch(sourceMapsConfig.include, filePath);
  const { debugMode } = getConfig();
  const generatorOptions: GeneratorOptions = {
    sourceMaps: true,
    sourceFileName: toPosix(path.relative(getConfig().rootDir, filePath)),
    sourceRoot: 'reboost:///',
    minified: !debugMode
  }

  const { code: generatedCode, map: generatedMap } = generate(ast, sourceMapsEnabled ? generatorOptions : undefined);
  let map;

  if (sourceMap && sourceMapsEnabled) {
    const merged = await mergeSourceMaps(sourceMap, generatedMap);
    map = getCompatibleSourceMap(merged);
  }

  return {
    code: generatedCode,
    map: map && JSON.stringify(map, null, getConfig().debugMode ? 2 : 0),
    imports,
    dependencies,
    error: errorOccurred
  }
}
