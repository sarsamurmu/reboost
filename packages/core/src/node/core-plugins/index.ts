import { JSONPlugin } from './json';
import { LoaderPlugin } from './loader';
import { ResolverPlugin } from './resolver';
import { CommonJSPlugin } from './commonjs';
import { CommonJSInteropPlugin } from './commonjs-interop';

export const CorePlugins = [
  JSONPlugin,
  LoaderPlugin,
  ResolverPlugin,
  CommonJSPlugin,
  CommonJSInteropPlugin
]
