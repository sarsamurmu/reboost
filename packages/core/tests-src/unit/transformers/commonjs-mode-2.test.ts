import { runTransformation } from 'src-node/core-plugins/commonjs-mode-2';
import { createTransformer } from './transformer';

const t = createTransformer((ast) => runTransformation(ast, ''));
const match = (code: string) => expect(t(code)).toMatchSnapshot();

describe('fixes <exportsObj>', () => {
  test('when dot notation is used', () => {
    match(`
      module.exports.item1 = 0;
      exports.item2 = 0;
    `);
  });

  test('when bracket notation is used', () => {
    match(`
      module.exports['item1'] = 0;
      module['exports'].item2 = 0;
      exports['item3'] = 0;
    `);
  });

  test('uses same identifier for same exported name', () => {
    match(`
      module.exports.item1 = 0;
      exports.item1 = 0;
    `);
  });
});

describe("does not fix <exportsObj> if <exportsObj>'s variable is already declared", () => {
  test('in the Program scope', () => {
    match(`
      const module = { exports: {} };
      module.exports.item1 = 0;
    `);

    match(`
      const exports = {};
      exports.item1 = 0;
    `);
  });

  test('in the current scope', () => {
    match(`
      module.exports.item1 = 0;
      {
        const module = { exports: {} };
        module.exports.item2 = 0;
      }
    `);

    match(`
      module.exports.item1 = 0;
      callback((module) => {
        module.exports.item2 = 0;
      });
    `);

    match(`
      exports.item1 = 0;
      {
        const exports = {};
        exports.item2 = 0;
      }
    `);

    match(`
      exports.item1 = 0;
      callback((exports) => {
        exports.item2 = 0;
      });
    `);
  });
});

describe('does not export default <exportsObj>', () => {
  test('when default export is available', () => {
    match(`
      module.exports.default = 0;
    `);

    match(`
      exports.default = 0;
    `);
  });

  describe('when the module is declared __esModule', () => {
    test('using Object.defineProperty', () => {
      match(`
        Object.defineProperty(module.exports, '__esModule', { value: true });
        module.exports.item1 = 0;
      `);

      match(`
        Object.defineProperty(exports, '__esModule', { value: true });
        exports.item1 = 0;
      `);
    });

    test('using <exportObj>.__esModule', () => {
      match(`
        module.exports.__esModule = true;
        module.exports.item1 = 0;
      `);

      match(`
        exports.__esModule = true;
        exports.item1 = 0;
      `);
    });
  });
});

test('fixes require calls', () => {
  match(`
    const module_1 = require('module_1');
    const module_1_1 = require('module_1');
  `);
});

describe('does not fix require calls', () => {
  test('if call signature is different', () => {
    match(`
      const result_1 = require('someThing', 0);
      const result_2 = require(100);
    `);
  });

  test('if it is built-in module', () => {
    match(`
      const fs = require('fs');
      const path = require('path');
    `);
  });

  describe('if "require" variable is already declared', () => {
    test('in Program scope', () => {
      match(`
        const require = (mod) => someFunc(mod);
        const module_1 = require('module_1');
      `);
    });

    test('in current scope', () => {
      match(`
        const module_1 = require('module_1');
        {
          const require = (mod) => someFunc(mod);
          const result = require('someThing');
        }
      `);

      match(`
        const module_1 = require('module_1');
        callback((require) => {
          const result = require('someThing');
        });
      `);
    });
  });
});

describe('fixes module.exports = require()', () => {
  test('exports default from the exported module if only one exports is available', () => {
    match(`
      module.exports = require('module_1');
    `);
  });

  test('transforms require call when other exports are available after the statement', () => {
    match(`
      module.exports = require('module_1');
      module.exports.item1 = 0;
      exports.item2 = 0;
    `);
  });

  test('resets previous exports', () => {
    match(`
      module.exports.item1 = 0;
      exports.item2 = 0;
      module.exports = require('module_1');
    `);
  });
});
