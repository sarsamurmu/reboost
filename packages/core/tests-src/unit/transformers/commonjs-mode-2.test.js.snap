// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not export default <exportsObj> when default export is available 1`] = `
"const module = {
  exports: {}
};
let export_0_;
module.exports.default = 0;
export_0_ = module.exports.default;
export { export_0_ as default };"
`;

exports[`does not export default <exportsObj> when default export is available 2`] = `
"const exports = {};
let export_0_;
exports.default = 0;
export_0_ = exports.default;
export { export_0_ as default };"
`;

exports[`does not export default <exportsObj> when the module is declared __esModule using <exportObj>.__esModule 1`] = `
"const module = {
  exports: {}
};
let export_0_;
module.exports.__esModule = true;
module.exports.item1 = 0;
export_0_ = module.exports.item1;
export { export_0_ as item1 };"
`;

exports[`does not export default <exportsObj> when the module is declared __esModule using <exportObj>.__esModule 2`] = `
"const exports = {};
let export_0_;
exports.__esModule = true;
exports.item1 = 0;
export_0_ = exports.item1;
export { export_0_ as item1 };"
`;

exports[`does not export default <exportsObj> when the module is declared __esModule using Object.defineProperty 1`] = `
"const module = {
  exports: {}
};
let export_0_;
module.exports.item1 = 0;
export_0_ = module.exports.item1;
export { export_0_ as item1 };"
`;

exports[`does not export default <exportsObj> when the module is declared __esModule using Object.defineProperty 2`] = `
"const exports = {};
let export_0_;
exports.item1 = 0;
export_0_ = exports.item1;
export { export_0_ as item1 };"
`;

exports[`does not fix <exportsObj> if <exportsObj>'s variable is already declared in the Program scope 1`] = `
"const module = {
  exports: {}
};
module.exports.item1 = 0;"
`;

exports[`does not fix <exportsObj> if <exportsObj>'s variable is already declared in the Program scope 2`] = `
"const exports = {};
exports.item1 = 0;"
`;

exports[`does not fix <exportsObj> if <exportsObj>'s variable is already declared in the current scope 1`] = `
"const module = {
  exports: {}
};
let export_0_;
module.exports.item1 = 0;
export_0_ = module.exports.item1;
{
  const module = {
    exports: {}
  };
  module.exports.item2 = 0;
}
export { export_0_ as item1 };
export default module.exports;
export const __cjsModule = true;"
`;

exports[`does not fix <exportsObj> if <exportsObj>'s variable is already declared in the current scope 2`] = `
"const module = {
  exports: {}
};
let export_0_;
module.exports.item1 = 0;
export_0_ = module.exports.item1;
callback(module => {
  module.exports.item2 = 0;
});
export { export_0_ as item1 };
export default module.exports;
export const __cjsModule = true;"
`;

exports[`does not fix <exportsObj> if <exportsObj>'s variable is already declared in the current scope 3`] = `
"const exports = {};
let export_0_;
exports.item1 = 0;
export_0_ = exports.item1;
{
  const exports = {};
  exports.item2 = 0;
}
export { export_0_ as item1 };
export default exports;
export const __cjsModule = true;"
`;

exports[`does not fix <exportsObj> if <exportsObj>'s variable is already declared in the current scope 4`] = `
"const exports = {};
let export_0_;
exports.item1 = 0;
export_0_ = exports.item1;
callback(exports => {
  exports.item2 = 0;
});
export { export_0_ as item1 };
export default exports;
export const __cjsModule = true;"
`;

exports[`does not fix require calls if "require" variable is already declared in Program scope 1`] = `
"const require = mod => someFunc(mod);

const module_1 = require('module_1');"
`;

exports[`does not fix require calls if "require" variable is already declared in current scope 1`] = `
"const __commonJS_ = mod => mod.__cjsModule ? mod[\\"default\\"] : mod;

import * as imported_0_ from \\"module_1\\";

const module_1 = __commonJS_(imported_0_);

{
  const require = mod => someFunc(mod);

  const result = require('someThing');
}"
`;

exports[`does not fix require calls if "require" variable is already declared in current scope 2`] = `
"const __commonJS_ = mod => mod.__cjsModule ? mod[\\"default\\"] : mod;

import * as imported_0_ from \\"module_1\\";

const module_1 = __commonJS_(imported_0_);

callback(require => {
  const result = require('someThing');
});"
`;

exports[`does not fix require calls if call signature is different 1`] = `
"const result_1 = require('someThing', 0);

const result_2 = require(100);"
`;

exports[`does not fix require calls if it is built-in module 1`] = `
"const fs = require('fs');

const path = require('path');"
`;

exports[`fixes <exportsObj> uses same identifier for same exported name 1`] = `
"const exports = {};
const module = {
  exports
};
let export_0_;
module.exports.item1 = 0;
export_0_ = module.exports.item1;
exports.item1 = 0;
export_0_ = exports.item1;
export { export_0_ as item1 };
export default module.exports;
export const __cjsModule = true;"
`;

exports[`fixes <exportsObj> when bracket notation is used 1`] = `
"const exports = {};
const module = {
  exports
};
let export_0_, export_1_;
module.exports['item1'] = 0;
export_0_ = module.exports.item1;
exports['item2'] = 0;
export_1_ = exports.item2;
export { export_0_ as item1, export_1_ as item2 };
export default module.exports;
export const __cjsModule = true;"
`;

exports[`fixes <exportsObj> when dot notation is used 1`] = `
"const exports = {};
const module = {
  exports
};
let export_0_, export_1_;
module.exports.item1 = 0;
export_0_ = module.exports.item1;
exports.item2 = 0;
export_1_ = exports.item2;
export { export_0_ as item1, export_1_ as item2 };
export default module.exports;
export const __cjsModule = true;"
`;

exports[`fixes module.exports = require() exports default from the exported module if only one exports is available 1`] = `
"import * as for_default_ from \\"module_1\\";
let default_export_;
export * from \\"module_1\\";
if (\\"default\\" in for_default_) default_export_ = for_default_[\\"default\\"];
export default default_export_;"
`;

exports[`fixes module.exports = require() resets previous exports 1`] = `
"import * as for_default_ from \\"module_1\\";
let default_export_;
const exports = {};
const module = {
  exports
};
module.exports.item1 = 0;
exports.item2 = 0;
export * from \\"module_1\\";
if (\\"default\\" in for_default_) default_export_ = for_default_[\\"default\\"];
export default default_export_;"
`;

exports[`fixes module.exports = require() transforms require call when other exports are available after the statement 1`] = `
"const __commonJS_ = mod => mod.__cjsModule ? mod[\\"default\\"] : mod;

import * as imported_0_ from \\"module_1\\";
const exports = {};
const module = {
  exports
};
let export_0_, export_1_;
module.exports = __commonJS_(imported_0_);
module.exports.item1 = 0;
export_0_ = module.exports.item1;
exports.item2 = 0;
export_1_ = exports.item2;
export { export_0_ as item1, export_1_ as item2 };
export * from \\"module_1\\";
export default module.exports;
export const __cjsModule = true;"
`;

exports[`fixes require calls 1`] = `
"const __commonJS_ = mod => mod.__cjsModule ? mod[\\"default\\"] : mod;

import * as imported_0_ from \\"module_1\\";

const module_1 = __commonJS_(imported_0_);

const module_1_1 = __commonJS_(imported_0_);"
`;
